[포커게임 앱 Step5]  미션 제출합니다.

ehgud0670 opened the pull request:
> 고민한 점

* segmentedControl 의 버튼이 눌릴때마다 뷰가 업데이팅 되도록 할 때 delegate 패턴을 써보도록 하였습니다.
* 각각의 StackView를 UIStackView를 상속받게 해 뷰컨트롤러에서 뷰에 대한  코드가 적어지도록 하였습니다.
* 처음에 setupView()로 stut=7, playersNum= 4로 세팅하고 updateView에서 주어지는 stut과 playersNum를 기준으로 해당되는 view들은 isHidden을 false로, 해당되지 않는 view들을 isHidden을 true로 하였습니다. 
* 매 게임마다 랜덤으로 카드가 셔플되도록 뷰컨트롤러에 카드덱을 프로퍼티로 두고 게임객체의 매개변수로 주었습니다.

ehgud0670 committed:
Refactor : init함수 지우고 reset() 함수로 변경
	* cards의 타입 non - optional 로 변경

ehgud0670 committed:
Refactor : 에러 처리 안하도록 코드변경 및 receiveCard 매개변수 타입 변경

	* Deck의 removeOne()의 throws를 없애고 반환값 Card?으로 변경
	(이유 : remove()에 의한 에러가 명확하고,
	다른 이유의 에러는 없기때문에 에러 처리 안하게 변경)

	* Player의 receiveCard의 파라미터를 Card로 변경
	(이유 : 이전의 클로저의 sender()을 호출할 필요없이 플레이어는
	카드만 받으면 되기 때문, 또 에러처리 관련 파라미터 코드도 삭제

	* 에러 처리를 없앤 코드에 따라 테스트 코드도 재작성

ehgud0670 committed:
Refactor : Players라는 클래스 만들어 코드를 단순화함

ehgud0670 committed:
Add : 카드 이미지 추가

ehgud0670 committed:
Refactor : 카드 파일명에 따라 description 수정

ehgud0670 committed:
Deleted : 현재 미션과 관련없은 이전의 미션 메서드들 삭제

ehgud0670 committed:
Feature : PlayersNum enum을 이용해 segmentControls의 items에 사용하려고 CustomStringConvertible 사용

ehgud0670 committed:
Feature : 이후에 사용할 segmentedControls를 담을 StackView 추가

ehgud0670 committed:
Deleted : 지금 미션에 필요없는 이전 미션의 CardStackView 삭제
	* 앞으로 구현하면서 필요한 때에 다시 CardStackView를 재작성할 예정

ehgud0670 committed:
Feature : stut과 players 관련한 segmentedControl 생성 및 stackView에 추가

ehgud0670 committed:
Refactor : UISegmentControl 상속한 클래스 생성해서 stackView에 추가

ehgud0670 committed:
Feature : segmentContorl의 segment 누르면 뷰컨트롤러가 print하도록 기능 구현
	* print는 앞으로의 기능 구현을 위한 임시용이라 삭제할것임

ehgud0670 committed:
Feature : 1줄에5장씩 2줄 카드뒷면이 보이도록 기능 구현

ehgud0670 committed:
Refactor : SegmentedControl 버튼 한번 눌렀을때 뷰컨트롤러에 잘 적용되도록 코드 수정

ehgud0670 committed:
Refactor: lazy var가 안되도록 코드 수정

ehgud0670 committed:
Feature: 뷰가 기본 세팅(카드:7, 참여자:5)이 되도록 기능 구현

ehgud0670 committed:
Refactor : 외부에 안보이도록 private 선언 및 위치변경

ehgud0670 committed:
Feature: playersNum 프로퍼티 추가 및 enum 이름 Number로 변경

ehgud0670 committed:
Feature : 초기에 stut 7, players 1로 게임시작되고 뷰가 반영되도록 기능 구현
	* 각 뷰에 updateView 메소드 기능 구현

ehgud0670 committed:
Refactor : 라벨 보이도록 기능 수정 및 가독성을 위해 메소드 여러개로 나눔

ehgud0670 committed:
Style : 일관성을 위해 파라미터 위치들 변경

ehgud0670 committed:
Feature : segmentControl의 input에 따라 뷰가 다르게 보이도록 기능 구현

ehgud0670 committed:
Refactor : 뷰 업데이트할때 먼저 보이도록하고 이후에 안보이도록 기능 수정
	* isHidden = false를 전체적으로 먼저 하고
	* isHidden = true를 범위밖에 subview만 적용되도록 기능 수정
	이유 : 이전의 순서는 라벨이 updating이 되지 않아서 이렇게
	코드 수정 함

ehgud0670 committed:
Refacotor : 참여자들의 카드 객체가 뷰에 매핑되도록 기능 구현

ehgud0670 committed:
Feature : Shake 이벤트를 하면 랜덤 카드를 다시 섞고 다시 동작하도록 구현

ehgud0670 committed:
Refactor: 게임 할 때마다 셔플 되도록 뷰컨트롤러에 카드덱 객체 생성
          * 카드덱 객체 게임 할때 마다 셔플되게 함
          * 게임 객체 만들때 카드덱 객체 매개변수로 넘겨줌
          * 게임 객체 생성시 hasEnoughCards()를 이용해
             카드 덱의 카드 개수가 부족하면 다시 덱을

ehgud0670 committed:
Refactor : 레이블 텍스트 Players 에서 Player로 되도록 코드 수정

ehgud0670 committed:
Modified : step5에 맞게 README.md 파일 작성

godrm commented on CardGameApp/CardGameApp/Model/Players.swift:
```diff
@@ -0,0 +1,51 @@
+
+    func searchPlayersNum(handler: (Number) -> ()) {
+        handler(playersNum)
+    }
```
searchPlayersNum 이건 왜 필요한 건지 잘 모르겠습니다.
Number를 넘겨줘야 하는 경우가 있다면 그냥 return 해주는 타입이 있어도 되지 않나요?
모든 경우를 클로저로 만들어야 하는 것은 아닙니다.

ehgud0670 replied:
솔직히 get,set 방식은 하면 안되는 방식이고 무조건 클로져를 해야한다고 느껴져서  객체안의 프로퍼티 값이 필요할 때 모두 클로져를 만들었습니다. 
get, set 방식을 써도 되는 상황을 잘 모르겠습니다. 지혜를 주십쇼.. ㅠㅠ

godrm replied:
어제 얘기했던 내용인거죠? 
우선 로직을 상위 모듈에서 구현해야만 하는 경우 (클로저 활용), 로직을 하위 모듈에서 구현하고 결과만 받아야 하는 경우(return 타입 활용) 정도만 나눠도 좋을 것 같습니다.
그외에 필요하면 getter, setter를 쓰는건데, 이걸 우선 고려하지 말자는 겁니다.

ehgud0670 replied:
오 넵 감사합니다!👍👍👍👍

godrm commented on CardGameApp/CardGameApp/Model/Players.swift:
```diff
@@ -0,0 +1,51 @@
+    var description: String {
+        let num = String(self.rawValue)
+        let people = "명"
+        return num + people
```
Number 같은 타입이 값을 주는 것과 출력할 형식을 만드는 것을 고민해 볼 필요는 있습니다.
왜냐하면 `명`이 붙을지 안 붙을지는 Number가 모르고, 어떤 뷰에 어떤 형식으로 표시하느냐에 따라 달라지기 때문입니다.
그래서 값을 표현하는 타입과 그것을 출력하는 것을 분리하는 게 좋습니다. 참고하세요.

ehgud0670 replied:
오 감사합니다. 분리해보겠습니다.

godrm commented on CardGameApp/CardGameApp/Model/PokerGame.swift:
```diff
@@ -0,0 +1,139 @@
+                playersNumCount += 1
+            }
+        }
+        return playersNumCount
```
위에도 남겼지만 이런 부분까지 클로저로 만들어야 하는 건 아닙니다. 
그냥 간단하게 그냥 값을 넘겨야 하는 건 넘겨도 됩니다. 그걸 선택하는 기준을 잡아보세요.

ehgud0670 replied:
리턴값이 여러개 있는 경우 클로져를 사용하고, 리턴값이 하나만 있는 경우는 그냥 값을 넘기면 될까요??

godrm commented on CardGameApp/CardGameApp/Model/PokerGame.swift:
```diff
@@ -0,0 +1,139 @@
+    func searchPlayers(handler: (Player) -> ()) {
+        players.searchPlayer {
+            handler($0)
+        }
```
여기 searchDealer와 searchPlayer는 동일한 인터페이스를 갖고 있는데 한꺼번에 처리할 방법이 있었는지 한 번 생각해보세요. (개선하자는 게 아니라 확인해보세요)

ehgud0670 replied:
생각해보겠습니다..

ehgud0670 replied:
```swift
   func searchPariticipants(playersHandler: (Player) -> (), dealerHandler: (Player) -> ()) {
        players.searchPlayer {
            playersHandler($0)
        }
        dealerHandler(dealer)
    }
```

이런식으로 각각의 다른 핸들러에 값을 넘겨주면 되지 않을 까요?? 궁금합니다!

godrm replied:
이것도 하나의 방법이겠네요. 두 개를 단순히 합치는 방법 뿐일까요. 
아니면 인터페이스가 동일하다면 딜러를 player 배열과 임시로 합쳐서 다 호출하는 건 어떨까요
선택지가 있으면 기준을 정해야 결정할 수 있습니다. 
다양한 방법을 찾아서 결정하는 게 엔지니어의 역할이에요.

godrm commented on CardGameApp/CardGameApp/View/CardsStackView.swift:
```diff
@@ -0,0 +1,86 @@
+    }
+    
+    required init(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
```
커스텀 뷰를 상속받아 만들 때는 init() 부분을 고려해야 합니다. 
init(frame:)과 init(coder:) 어떻게 다른가요? 어느 한 쪽만 구현해야 할까요? 
두 개가 모두 동일하게 동작하도록 하는게 좋습니다.

ehgud0670 replied:
* init(frame:)은 초기 좌표를 superview에게 전달함으로써 객체를 만드는 방식입니다. 
autolayout을 사용할 경우 autolayout의 contraints가 superview의 최종 크기와 배치를 결정하므로
이런 경우에는 superview에게 CGRectZero를 넘겨주면 됩니다. `super.init(frame: .zero)`

* init(coder:)의 경우 NSCoder가 unarchiver 객체, 즉 decoder인데 decoder로 아카이브파일을  decode해서 나온  **data**로 객체를 만드는 방식입니다. 
만들어지는 객체가 storyboard를 바탕으로 객체라면, 즉 nib file에 있는 것을 객체화해야 되는 상황이라면 
init(coder:)를 사용해야 합니다.

두 경우 모두 동일하게 동작하도록 얼른 수정하고 push 하겠습니다!

godrm commented on CardGameApp/CardGameApp/View/CardsStackView.swift:
```diff
@@ -0,0 +1,86 @@
+        
+        for _ in 0 ..< maxNum {
+            addArrangedSubview(generateCardImageViewDefault())
+        }
```
위에 forEach가 7번 동작하고 아래 for 문이 7번 동작하지 않나요?
그냥 위에서 처리하면 되는거 아닌가요?

ehgud0670 replied:
오 그렇네요, 피드백 감사합니다. 👍

godrm commented on CardGameApp/CardGameApp/View/GameSegmentedControlStackView.swift:
```diff
@@ -0,0 +1,94 @@
+            return PokerGame.GameStut.five
+        default:
+            return nil
+        }
```
enum 타입을 사용할 때는 switch-case를 사용하기 보다는 init()을 활용하는 게 좋습니다. 
그렇지 않으면 enum 타입 값이 바뀔 때마다, selectedSegment 순서가 바뀔때마다 다시 다 바꿔야 합니다.

godrm commented on CardGameApp/CardGameApp/View/ParticipantsStackView.swift:
```diff
@@ -0,0 +1,87 @@
+    
+    private func updateDealer(playersCount: Int, dealer: Player) {
+        let dealerCount = 1
+        let dealerIndex = playersCount + dealerCount - 1
```
매번 players 와 dealer 개수를 더하고 빼고 하는 것보다는 고정하거나 별도로 처리하는 건 어떤가요? 
Player와 Dealer 객체가 다형성을 고려해보면 좋았을 것 같습니다.

ehgud0670 replied:
네 다형성을 고려하여 코드를 수정해보겠습니다. 👍

ehgud0670 replied:
다형성을 고려하라는게 예를 들어
Participant라는  super 클래스를 Player클래스와 Dealer클래스가 상속하는 방식으로 코드를 작성하라는 말씀인가요?
궁금합니다!!

godrm replied:
어느 쪽이 수퍼클래스가 되거나 공통의 수퍼 클래스가 될 수도 있겠죠

ehgud0670 replied:
```swift
extension PokerGame.PlayersNum: Comparable {
    static func < (lhs: PokerGame.PlayersNum, rhs: PokerGame.PlayersNum) -> Bool {
        return lhs.rawValue < rhs.rawValue
    }
    
    static func forEachMaxParticipantsCase(handler: () -> (Void)) {
        if let maxNumber = PokerGame.PlayersNum.allCases.max() {
            let dealerCount = 1
            let maxParticipantsNum = maxNumber.rawValue + dealerCount
            for _ in 0 ..< maxParticipantsNum {
                handler()
            }
        }
    }
    
    func forEachParticipantsCase(handler: () -> (Void)) {
        let dealerCount = 1
        let participantsNum = self.rawValue + dealerCount
        for _ in 0 ..< participantsNum {
            handler()
        }
    }
}
```
이렇게 enum에 forEachParticipantsCase()나 forEachMaxParticipantsCase() 를 만들어서 
그때그때 dealerCount를 더하지는 않는 방식으로 개선하기는 했는데 이게 다형성을 고려한건지는 잘 모르겠습니다..😔

godrm reviewed (commented):
다음 피드백을 확인해서 뷰를 만드는 부분에서 세부 사항을 개선해보세요.

ehgud0670 committed:
Refactor: 클로져안에서 addArrangedSubView()를 호출하도록 코드 수정
	이유 : 이전의 방법은 불필요하게 for문을 2번 호출하기 때문에
		하나로 합침

ehgud0670 committed:
Refactor: switch case 제거하고 대신 init으로 enum 리턴하게 코드 수정

ehgud0670 committed:
Refactor : 값을 표현하는 타입과 출력하는 부분을 분리함
	* 각 enum 의 CustomStringConvertible의 return 값은
	  enum들이 납득할 수 있는 값을 대체함
	* 피드백대로 '명'이 붙는지, 'Cards'가 붙는지는 enum인
	  GameStut과 Number가 몰라야 하기 때문에 이런식으로 코드 수정함

ehgud0670 committed:
Refactor: 클로저를 사용하는 것 대신 값을 넘겨주는 것으로 코드 수정
	* 값을 넘겨주기만 하면 되는 프로퍼티라 클로저를 쓰는 것은 삭제함

ehgud0670 committed:
Refactor: Players객체 대신 딜러도 포함할수 있는 Participants로 코드 수정
	* searchParticipants 메소드를 만들어 배열 players에서 임시로 dealer를 더해
	한번에 탐색할 수 있도록 함
	* update할때 dealer를 더하고 빼는 코드는 없도록 이렇게 코드 수정함

ehgud0670 committed:
Deleted: 불필요하다고 판단해 playersNum 프로퍼티 삭제
	이유 : playersNum enum은 particpants의 players 배열을 채우기 위한
	      용도로만 생각하기 때문이다.

ehgud0670 committed:
Refactor: PlayersNum enum을 PokerGame의 nested enum으로 변경
	 이유 : 이 enum이 Pokergame의 초기화를 위한 enum이라고 생각하기
	때문에 Participants의 enum이라기 보단 PokerGame의 enum이라고 생각
	하기 때문이다.

ehgud0670 committed:
Refactor : dealer가 항상 1개 있는 상황을 고려하여 ParticipantsNum이라는 struct  구현 및 적용
	이유 : update할때 마다 항상 dealerCount를 더하고 빼는 상황을 없애기 위해서 기능 구현함

ehgud0670 committed:
Refactor: 이전 커밋에서 만든 forEachMaxCase()를 setupView()에 적용
	이유 : dealerCount를 1 더하는 상황을 없애기 위해서

ehgud0670 committed:
Refactor: enum GameStut을 PokerGame으로 분리함
	이유1 : 미션을 하면 할수록 GameStut에 대한 코드가 늘어남
	따라서 GameStut에 대한 코드가 늘어나면
	PokerGame.swift 파일이 더 읽기 어려워지므로 아예 분리함
	이유2 : PokerGame을 통해 GameStut을 사용하는 것이 코드의
	가독성을 해친다고 생각하기 때문이다.
	before : PokerGame.GameStut
	after : GameStut

ehgud0670 committed:
Refactor: struct ParticipantsNum을 PokerGame으로 분리함
            이유1 : 미션을 하면 할수록 ParticipantsNum에 대한 코드가 늘어남
            따라서 ParticipantsNum에 대한 코드가 늘어나면
            PokerGame.swift 파일이 더 읽기 어려워지므로 아예 분리함
            이유2 : PokerGame을 통해 ParticipantsNum을 사용하는 것이 코드의
            가독성을 해친다고 생각하기 때문이다.
            before : PokerGame.ParticipantsNum
            after : ParticipantsNum

ehgud0670 committed:
Refactor: Participants 클래스에 프로퍼티 name을 생성 및 view update할 때 적용
	이유1: view update할 때 매번 Players + number로 name을 update하는 것을 피하기 위해
	이유2: 이전의 코드는 dealer의 name을 넣지 못하는데 dealer의 name을 넣기 위해

ehgud0670 committed:
Style: 타입 지정하는 곳의 공백 지움

ehgud0670 committed:
Refactor: 가독성을 위해 메소드 분리함

ehgud0670 committed:
Style: player보다 participant가 더 합당한 이름이라 이름 변경함

ehgud0670 committed:
Feature : init(coder:)의 경우도 객체가 정상적으로 만들어지도록 기능 구현
	이유: init(coder:)는 storyboard에 있는 뷰를 쓰는 경우에 호출되는 init()인데
	이후에 storyboard로 만들수도 있는 있는 가능성 때문에
	이 경우에도 custom view가 정상적으로 초기화 및 설정되도록
	하기 위해 기능 구현함

ehgud0670 commented:
* 피드백 고친 것들 
1. 클로져가 불필요하다고 생각된 부분은 값을 넘겨주는 것으로 수정하였습니다.  
2. enum을 정의하는 부분과 출력하는 부분을 분리하였습니다. enum의 description은 String(self.rawValue)를 리턴하고 뷰 쪽에서 ‘명’이나 ‘cards’을 붙였습니다.
3. 뷰 쪽에서 update 할 때  dealerCount=1을 매번 더해야 하는 구조였는데 participantsNum이라는 구조체를 만들어서 숫자가 고정되게끔 수정하였습니다. 
4. searchPlayers와 searchDealer가 같은 인터페이스를 갖고 있는데 합치는것도 생각해보라는 피드백을 받았고 participants 라는 객체를 만들어서 searchParticipants()를 만들어 해결하였습니다.
6. segmentedControl의 selectedIndex와 관련하여 불필요한 switch case를 제거하고 enum의 allCase를 활용하여 index로 접근하는 것으로 코드를 수정하였습니다. 
7. init(coder:)가 언제 필요한지 알아보고 init(frame:)과 정상 작동하도록 코드를 수정하였습니다. 

> 새로 알게 된 점

다형성을 고려하라는 말씀을 듣고 어제 버스타고 귀가길에 계속 고민해보다가 아이디어가 하나 떠올랐습니다. 
애초에 딜러인 dealer 변수와 플레이어들을 나타내는 players를 따로 프로퍼티로 두는게 아니라 dealer 가 고정이니  Participants라는 클래스를 만들고 프로퍼티로 
Players와 dealer를 두었습니다. 이러면 searchParticipants에 모든 참여자들을 탐색할 수 있는 장점이 있었습니다. 
또 participant 객체에 name이라는 프로퍼티를 두어 dealer와 players간에 구별할 수 있게 두었습니다.
이후에 dealer와 player 각각에 따라 기능이 다른 점이 있다면 상속을 통해 해결하려고도 생각하고 있습니다. 

* 사고의 전환이 제일 중요한 것 같습니다! 👍 

> 아쉬운점 

Participants 객체가 있고, ParticipantsNum 구조체가 또 따로 있는게 마음에 걸립니다. 만약 dealer가 2명으로 된다면 Participants나 ParticipantsNum 모두를 수정해야하는 데 이상적인 방법이 아니라고 생각하기 때문입니다. 하나로 합칠 수 없는지… 고민되네요 😅

godrm commented on CardGameApp/CardGameApp/Model/ParticipantsNum.swift:
```diff
@@ -0,0 +1,53 @@
+    var description: String {
+        return String(self.rawValue)
+    }
+}
```
네 얘기한데로 이 객체가 분리된 이유가 명확하지 않은 것 같습니다. 이 부분은 따로 좀 더 생각해보세요.

ehgud0670 replied:
네엡

godrm commented:
전체적으로 개선 사항이 많았습니다. 다음 단계를 진행하세요.
